File: src/bumperbot_py_examples/bumperbot_py_examples/simple_parameter.py
import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter

class SimpleParameter(Node):
    def __init__(self):
        super().__init__("simple_parameter_node")
        # Declare a parameter with a default value
        self.declare_parameter("simple_init_param", 28)
        self.declare_parameter("simple_string_param", "Hello,Fouad")
        
        self.add_on_set_parameters_callback(self.param_change_callback)

    def param_change_callback(self, params):
        result = SetParametersResult()
        for param in params:
            if param.name == "simple_init_param" and param.type_ == Parameter.Type.INTEGER:
                self.get_logger().info(f"Parameter 'simple_int_param' changed to: {param.value}")
                result.successful = True
            elif param.name == "simple_string_param" and param.type_ == Parameter.Type.STRING:
                self.get_logger().info(f"Parameter 'simple_string_param' changed to: {param.value}")
                result.successful = True
            else:
                result.successful = False
                result.reason = f"Parameter '{param.name}' change rejected."
        return result

def main():
    rclpy.init()
    simple_param_node = SimpleParameter()
    rclpy.spin(simple_param_node)
    simple_param_node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/bumperbot_cpp_examples/src/simple_parameter.cpp
#include <rclcpp/rclcpp.hpp>
#include <string>
#include <vector>
#include <rcl_interfaces/msg/set_parameters_result.hpp>
using std::placeholders::_1;
class SimpleParameterNode : public rclcpp::Node
{
public:
    SimpleParameterNode(): Node("simple_parameter_node")
    {
        // Declare a parameter with a default value
        this->declare_parameter<int>("simple_int_param", 28);
        this->declare_parameter<std::string>("simple_string_param", "Hello, ROS2!");

        param_callback_handle_ = add_on_set_parameters_callback(std::bind(&SimpleParameterNode::paramChangeCallback, this, _1));
        // Log the parameter values
        RCLCPP_INFO(this->get_logger(), "Integer parameter: %d", this->get_parameter("simple_int_param").as_int());
        RCLCPP_INFO(this->get_logger(), "String parameter: %s", this->get_parameter("simple_string_param").as_string().c_str());
    }

private:
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    rcl_interfaces::msg::SetParametersResult paramChangeCallback(const std::vector<rclcpp::Parameter> & parameters)
    {
        rcl_interfaces::msg::SetParametersResult result;
        for (const auto &param : parameters)
        {
            if (param.get_name() == "simple_int_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER)
            {
                int new_value = param.as_int();
                RCLCPP_INFO(this->get_logger(), "simple_int_param changed to: %d", new_value);
                result.successful = true;
            }
            else if (param.get_name() == "simple_string_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_STRING)
            {
                std::string new_value = param.as_string();
                RCLCPP_INFO(this->get_logger(), "simple_string_param changed to: %s", new_value.c_str());
                result.successful = true;
            }
            else
            {
                result.successful = false;
                result.reason = "Invalid parameter or type";
                RCLCPP_WARN(this->get_logger(), "Failed to change parameter: %s", param.get_name().c_str());
                return result;
            }
        }
        return result;
    }
};

int main (int argc, char ** argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleParameterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

